### ISSUE TITLE
py-test import failure

### ISSUE THREAD

**Issue by gdt** [pyproj4/pyproj#1447]
# Problem description

I am updating the pkgsrc pkg for py-proj to 3.7.0.   I do not have a clear memory of running tests before.  In pkgsrc we invoked tests as 'make test' from the package directory, which invokes tests in a build-system-specific way in the build directory.  This generally works, and it works with py-test which is normal.

Many packages load resources from the system rather than the build directory, making them not testable before install.  As I generally don't try to fix those things, my normal practice is to install the new bits first and then test, so that test code that loads pyproj from /usr/pkg/lib/python3.12/site-packages/pyproj instead of the build dir will run as expected.

Then, I ran tests, and it failed to find _context.  But, running the test program itself produced no errors, and I do see _context.so in the site-lib dir.  ldd shows all its libs resolve ok.

Maybe it is trying to use the build tree?  My build tree has _context.so and it looks normal:
```
> find . -name _context\*
./pyproj/_context.pxd
./pyproj/_context.pyi
./pyproj/_context.pyx
./pyproj/_context.c
./build/lib.netbsd-10.0_STABLE-amd64-cpython-312/pyproj/_context.pxd
./build/lib.netbsd-10.0_STABLE-amd64-cpython-312/pyproj/_context.pyi
./build/lib.netbsd-10.0_STABLE-amd64-cpython-312/pyproj/_context.pyx
./build/lib.netbsd-10.0_STABLE-amd64-cpython-312/pyproj/_context.so
./build/temp.netbsd-10.0_STABLE-amd64-cpython-312/pyproj/_context.o
```

```
$ pytest-3.12 
ImportError while loading conftest '/tmp/work/geography/py-proj/work/pyproj-3.7.0/test/conftest.py'.
test/conftest.py:10: in <module>
    import pyproj
pyproj/__init__.py:34: in <module>
    import pyproj.network
pyproj/network.py:10: in <module>
    from pyproj._context import _set_context_ca_bundle_path
E   ModuleNotFoundError: No module named 'pyproj._context'
$ python3.12 test/conftest.py 
```
# Environment Information

```
pyproj info:
    pyproj: 3.7.0
PROJ (runtime): 9.5.0
PROJ (compiled): 9.5.0
  data dir: /usr/pkg/share/proj
user_data_dir: /home/gdt/.local/share/proj
PROJ DATA (recommended version): 1.19
PROJ Database: 1.4
EPSG Database: v11.016 [2024-08-31]
ESRI Database: ArcGIS Pro 3.3 [2024-08-16]
IGNF Database: 3.1.0 [2019-05-24]

System:
    python: 3.12.6 (main, Sep 18 2024, 14:20:12) [GCC 10.5.0]
executable: /usr/pkg/bin/python3.12
   machine: NetBSD-10.0_STABLE-amd64-x86_64-64bit-ELF

Python deps:
   certifi: 2024.8.30
    Cython: 3.0.11
setuptools: 75.1.0
       pip: None
```

**Comment by jdkloe** [comment]
Yes, this must be related to testing inside the source code directory. I get a similar error if I install the source tarball using pip. After installing pyproj and copying the testfiles to an alternative location the testing works for me:
```
cd ..
mkdir pyproj-test-folder
cd pyproj-test-folder
cp -r ../pyproj-$THISVERSION/test .
cp ../pyproj-$THISVERSION/pytest.ini .
py.test-3
```

**Comment by djhoese** [comment]
There are two important pieces to this type of problem:

1. Compiling C extensions.
2. Running tests from the correct directory.

### Extensions

Pyproj includes C extensions (via Cython) that must be compiled to shared library (ex. `.so` files on linux) before it can be properly used. Binary wheel files have these compiled libraries available already. If starting from the source code, they must be compiled by installing the library. A normal `pip install pyproj` will do this, but you can also install pyproj in an "editable" way:

```bash
pip install -e .
```

Where "." is the root of the pyproj repository where pyproject.toml is. Optionally include the `--no-deps` flag so it doesn't install dependencies if you already have them. This will build the Cython modules and put `.so` files (again, different extensions for different OSes) in the source directory. This will make it possible to run the tests (see below) from the source directory directly.

### Running tests

Pyproj's tests are not installed or available via python import from the install package. They are only in the `sdist` tarball or in the source git repository (someone correct me if I'm wrong). Additionally, if you run python and do `import pyproj` from a directory that has a `pyproj/__init__.py` file available then it will attempt to import that local pyproj instead of any installed pyproj. So if you were hoping to run the tests on a previously installed pyproj then you must `cd` to another directory and call `pytest /path/to/pyproj_src/test`. No need to copy the test directory, you just need to be outside of the directory where `pyproj/__init__.py` is accessible.

If you installed pyproj in an editable way (see previous section) then this import situation should work as expected and you can be in the root of the source directory and run `pytest test/`.

Here's a simple test I did while in my `/tmp` directory:

```bash
$ conda create -y -n pyproj_tests python=3.13 pyproj numpy pytest
$ conda activate pyproj_tests
$ conda uninstall -y --force pyproj

$ pip install -e ~/repos/git/pyproj --no-deps
$ pytest ~/repos/git/pyproj/test/  
...
=== 981 passed, 25 skipped, 143 warnings in 49.98s ===
```

I'm going to close this issue now. Bottom line is that it is important to install pyproj in some way before running tests so that extensions are completely built/compiled.

Feel free to comment further, but for anyone facing a similar issue in the future please file a new issue.
